Add Support for the dotnetcore3.1 runtime

alex-vance opened the pull request:
- Added the dotnetcore3.1 lambda runtime

May resolve issue #2285

alex-vance commented on localstack/services/awslambda/lambda_executors.py:
```diff
@@ -733,7 +734,7 @@ def docker_image_for_runtime(cls, runtime):
         # TODO: remove prefix once execution issues are fixed with dotnetcore/python lambdas
         # See https://github.com/lambci/docker-lambda/pull/218
-        lambdas_to_add_prefix = ['dotnetcore', 'python2.7', 'python3.6', 'python3.7']
+        lambdas_to_add_prefix = ['dotnetcore2.0', 'dotnetcore2.1', 'python2.7', 'python3.6', 'python3.7']
```
FYI, it looks like the GitHub issue comment above this change has been fixed/closed. You might be able to remove this prefix now (the tags do exist based on what i saw, at least for the dotnetcore images). I thought about updating that, but figured it'd be best to separate that from this specific addition.

ot-alex-vance commented:
Also, I'm not a python developer, so I stumbled through this as best I could. I am unsure why the the error on the build is happening and see no mention of it anywhere in the code. Any help would be appreciated.

whummer commented on tests/integration/test_lambda.py:
```diff
@@ -755,8 +762,43 @@ def test_dotnet_lambda_running_in_docker(self):
+
+        # assert that logs are present
+        expected = ['Running .NET Core 3.1 Lambda']
+        self.check_lambda_logs(TEST_LAMBDA_NAME_DOTNETCORE31, expected_lines=expected)
```
Looks like this assertion is failing - we may have to make the `check_lambda_logs(..)` function a bit more resilient to different log formats:
```
======================================================================

FAIL: test_dotnet_lambda_running_in_docker (tests.integration.test_lambda.TestDotNetCore31Runtimes)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/localstack/localstack/tests/integration/test_lambda.py", line 801, in test_dotnet_lambda_running_in_docker

    self.check_lambda_logs(TEST_LAMBDA_NAME_DOTNETCORE31, expected_lines=expected)

  File "/home/travis/build/localstack/localstack/tests/integration/test_lambda.py", line 94, in check_lambda_logs

    self.assertIn(line, log_messages)

AssertionError: 'Running .NET Core 3.1 Lambda' not found in ['\x1b[32mSTART RequestId: 4426511a-e558-19e6-b969-b8464f027c18 Version: $LATEST\x1b[0m', 'Running .NET Core 3.1 Lambda\x1b[32mEND RequestId: 4426511a-e558-19e6-b969-b8464f027c18\x1b[0m', '\x1b[32mREPORT RequestId: 4426511a-e558-19e6-b969-b8464f027c18\tInit Duration: 138.22 ms\tDuration: 16.37 ms\tBilled Duration: 100 ms\tMemory Size: 1536 MB\tMax Memory Used: 39 MB\t\x1b[0m']
```

ot-alex-vance replied:
I saw that and tried a few different things. is that something that I should do as part of this PR or comment it out and add a TODO (since the log itself is technically working).

alex-vance replied:
I saw that and tried a few different things. is that something that I should do as part of this PR or comment it out and add a TODO (since the log itself is technically working).

whummer replied:
Feel free to comment it out and add a TODO - we can pick it up in the next iteration 👍

ot-alex-vance commented on tests/integration/lambdas/dotnetcore31/test/dotnetcore31.Tests/FunctionTest.cs:
```diff
@@ -0,0 +1,28 @@
+using System;
```
It's not required, it's just the standard `dotnet new` template. I'll remove it.

alex-vance replied:
I'll remove it.

alex-vance commented on tests/integration/lambdas/dotnetcore31/src/dotnetcore31/aws-lambda-tools-defaults.json:
```diff
@@ -0,0 +1,16 @@
+{
```
I'm fairly certain it can be removed, I'll give it a shot.

alex-vance commented on tests/integration/test_lambda.py:
```diff
@@ -755,8 +762,43 @@ def test_dotnet_lambda_running_in_docker(self):
-        testutil.delete_lambda_function(TEST_LAMBDA_NAME_DOTNETCORE2)
+        testutil.delete_lambda_function(TEST_LAMBDA_NAME_DOTNETCORE31)
+
+    def test_dotnet_lambda_running_in_docker(self):
```
I'll give it a shot, where would that method belong? Under testutil?

alex-vance replied:
I'll give that a shot.

alex-vance commented:
@whummer - i've updated the tests and removed the items you've asked for. Due to my lack of experience with Python, i had trouble creating a base class and allow the `@classmethod` declarations to be reused as well. I've found a way to just use a single common method on a dotnetcore runtimes test class that defines a test method for each runtime. Let me know if that isn't sufficient enough, but I may need some assistance in any changes you're looking for. Hope this helps!

whummer commented:
Thanks a lot @alex-vance for updating the PR - looks like there is only a small test issue remaining, related to deletion of one of the Test functions:
```
======================================================================
ERROR: test suite for <class 'tests.integration.test_lambda.TestDotNetCoreRuntimes'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/nose/suite.py", line 229, in run
    self.tearDown()
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/nose/suite.py", line 352, in tearDown
    self.teardownContext(ancestor)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/nose/suite.py", line 368, in teardownContext
    try_run(context, names)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/opt/code/localstack/tests/integration/test_lambda.py", line 738, in tearDownClass
    testutil.delete_lambda_function(TEST_LAMBDA_NAME_DOTNETCORE2)
  File "/opt/code/localstack/localstack/utils/testutil.py", line 85, in delete_lambda_function
    client.delete_function(FunctionName=name)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 316, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 626, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the DeleteFunction operation: Function not found: arn:aws:lambda:us-east-1:000000000000:function:test_lambda_dotnetcore2
-------------------- >> begin captured logging << --------------------
localstack.services.awslambda.lambda_api: WARNING: Function not found: arn:aws:lambda:us-east-1:000000000000:function:test_lambda_dotnetcore2
--------------------- >> end captured logging << ---------------------
```

alex-vance commented:
@whummer Looks like were finally good to go!

whummer commented:
Thanks again!
